core/src/main/java/ai/timefold/solver/core/impl/heuristic/selector/move/generic/list/ruin/ListRuinRecreateMove.java

```diff
@@ -0,0 +1,115 @@
+package ai.timefold.solver.core.impl.heuristic.selector.move.generic.list.ruin;
+
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.NavigableSet;
+import java.util.Set;
+import java.util.TreeSet;
+
+import ai.timefold.solver.core.api.score.director.ScoreDirector;
+import ai.timefold.solver.core.impl.domain.variable.ListVariableStateSupply;
+import ai.timefold.solver.core.impl.heuristic.move.AbstractMove;
+import ai.timefold.solver.core.impl.heuristic.move.Move;
+import ai.timefold.solver.core.impl.heuristic.selector.list.LocationInList;
+import ai.timefold.solver.core.impl.heuristic.selector.move.generic.RuinRecreateConstructionHeuristicPhase.RuinRecreateConstructionHeuristicPhaseBuilder;
+import ai.timefold.solver.core.impl.score.director.VariableDescriptorAwareScoreDirector;
+import ai.timefold.solver.core.impl.solver.scope.SolverScope;
+import ai.timefold.solver.core.impl.util.CollectionUtils;
+
+public final class ListRuinRecreateMove<Solution_> extends AbstractMove<Solution_> {
+
+    private final ListVariableStateSupply<Solution_> listVariableStateSupply;
+    private final List<Object> ruinedValueList;
+    private final Set<Object> affectedEntitySet;
+    private final RuinRecreateConstructionHeuristicPhaseBuilder<Solution_> constructionHeuristicPhaseBuilder;
+    private final SolverScope<Solution_> solverScope;
+    private final Map<Object, NavigableSet<RuinedLocation>> entityToOriginalPositionMap;
+    private final Map<Object, NavigableSet<RuinedLocation>> entityToNewPositionMap;
+
+    public ListRuinRecreateMove(ListVariableStateSupply<Solution_> listVariableStateSupply,
+            RuinRecreateConstructionHeuristicPhaseBuilder<Solution_> constructionHeuristicPhaseBuilder,
+            SolverScope<Solution_> solverScope, List<Object> ruinedValueList, Set<Object> affectedEntitySet) {
+        this.listVariableStateSupply = listVariableStateSupply;
+        this.constructionHeuristicPhaseBuilder = constructionHeuristicPhaseBuilder;
+        this.solverScope = solverScope;
+        this.ruinedValueList = ruinedValueList;
+        this.affectedEntitySet = affectedEntitySet;
+        this.entityToOriginalPositionMap = CollectionUtils.newIdentityHashMap(affectedEntitySet.size());
+        this.entityToNewPositionMap = CollectionUtils.newIdentityHashMap(affectedEntitySet.size());
+    }
+
+    @Override
+    protected Move<Solution_> createUndoMove(ScoreDirector<Solution_> scoreDirector) {
+        return new ListRuinRecreateUndoMove<>(this, listVariableStateSupply.getSourceVariableDescriptor(),
+                entityToNewPositionMap, entityToOriginalPositionMap);
```

Comment:
I suggest updating the logic from `ListRuinRecreateUndoMove` to make it consistent with the previous comment.